--- busybox-1.15.1/archival/cpio.c	2009-09-12 17:55:36.000000000 +0200
+++ busybox-1.15.1/archival/cpio.c	2010-07-16 19:53:14.368794345 +0200
@@ -422,7 +422,7 @@ int cpio_main(int argc UNUSED_PARAM, cha
 	if (archive_handle->ah_priv[2] != (void*) ~(ptrdiff_t)0
 	 && !(opt & CPIO_OPT_QUIET)
 	)
-		printf("%lu blocks\n", (unsigned long)(ptrdiff_t)(archive_handle->ah_priv[2]));
+		fprintf(stderr, "%lu blocks\n", (unsigned long)(ptrdiff_t)(archive_handle->ah_priv[2]));
 
 	return EXIT_SUCCESS;
 }
