diff -urpN busybox-1.15.1.orig/include/platform.h busybox-1.15.1/include/platform.h
--- busybox-1.15.1.orig/include/platform.h	2009-09-12 17:55:36.000000000 +0200
+++ busybox-1.15.1/include/platform.h	2013-09-09 19:34:04.142273357 +0200
@@ -265,14 +265,17 @@ typedef unsigned smalluint;
 # include <stdbool.h>
 #endif
 
-/* Try to defeat gcc's alignment of "char message[]"-like data */
-#if 1 /* if needed: !defined(arch1) && !defined(arch2) */
+/* Useful for defeating gcc's alignment of "char message[]"-like data */
+#if !defined(__s390__)
+    /* on s390[x], non-word-aligned data accesses require larger code */
 # define ALIGN1 __attribute__((aligned(1)))
 # define ALIGN2 __attribute__((aligned(2)))
+# define ALIGN4 __attribute__((aligned(4)))
 #else
 /* Arches which MUST have 2 or 4 byte alignment for everything are here */
 # define ALIGN1
 # define ALIGN2
+# define ALIGN4
 #endif
 
 
