--- busybox-1.15.1/shell/hush.c	2012-03-26 16:50:30.716661392 +0200
+++ busybox-1.15.1.hush/shell/hush.c	2012-03-26 16:50:00.557670368 +0200
@@ -3967,6 +3967,13 @@
 			argv_expanded = expand_strvec_to_strvec(argv + command->assignment_cnt);
 		}
 
+		/* if someone gives us an empty string: `cmd with empty output` */
+		if (!argv_expanded[0]) {
+			free(argv_expanded);
+			debug_leave();
+			return G.last_exitcode;
+		}
+
 		x = find_builtin(argv_expanded[0]);
 #if ENABLE_HUSH_FUNCTIONS
 		funcp = NULL;
