diff -urpN busybox-1.15.1.orig/editors/awk.c busybox-1.15.1/editors/awk.c
--- busybox-1.15.1.orig/editors/awk.c	2009-09-12 17:55:58.000000000 +0200
+++ busybox-1.15.1/editors/awk.c	2010-11-30 12:41:54.305809835 +0100
@@ -683,14 +683,27 @@ static ALWAYS_INLINE int isalnum_(int c)
 
 static double my_strtod(char **pp)
 {
+	char *cp = *pp;
 #if ENABLE_DESKTOP
-	if ((*pp)[0] == '0'
-	 && ((((*pp)[1] | 0x20) == 'x') || isdigit((*pp)[1]))
-	) {
-		return strtoull(*pp, pp, 0);
+	if (cp[0] == '0') {
+		/* Might be hex or octal integer: 0x123abc or 07777 */
+		char c = (cp[1] | 0x20);
+		if (c == 'x' || isdigit(cp[1])) {
+			unsigned long long ull = strtoull(cp, pp, 0);
+			if (c == 'x')
+				return ull;
+			c = **pp;
+			if (!isdigit(c) && c != '.')
+				return ull;
+			/* else: it may be a floating number. Examples:
+			 * 09.123 (*pp points to '9')
+			 * 00.123 (*pp points to '.')
+			 * fall through to strtod.
+			 */
+		}
 	}
 #endif
-	return strtod(*pp, pp);
+	return strtod(cp, pp);
 }
 
 /* -------- working with variables (set/get/copy/etc) -------- */
