diff -urpN busybox-1.15.1.org/findutils/grep.c busybox-1.15.1/findutils/grep.c
--- busybox-1.15.1.org/findutils/grep.c	2009-09-12 17:55:36.000000000 +0200
+++ busybox-1.15.1/findutils/grep.c	2011-11-24 11:56:08.600013664 +0100
@@ -255,7 +255,10 @@ static int grep_file(FILE *file)
 		while (pattern_ptr) {
 			gl = (grep_list_data_t *)pattern_ptr->data;
 			if (FGREP_FLAG) {
-				found |= (strstr(line, gl->pattern) != NULL);
+				found |= (((option_mask32 & OPT_i)
+					? strcasestr(line, gl->pattern)
+					: strstr(line, gl->pattern)
+					) != NULL);
 			} else {
 				if (!(gl->flg_mem_alocated_compiled & COMPILED)) {
 					gl->flg_mem_alocated_compiled |= COMPILED;
