diff -urpN uClibc-0.9.30.1.org/libc/inet/rpc/pm_getport.c uClibc-0.9.30.1/libc/inet/rpc/pm_getport.c
--- busybox-1.15.1/uClibc-0.9.30.1.org/libc/inet/rpc/pm_getport.c	2006-06-19 03:16:44.000000000 +0200
+++ busybox-1.15.1/uClibc-0.9.30.1/libc/inet/rpc/pm_getport.c	2013-10-21 14:33:45.002827043 +0200
@@ -73,7 +73,19 @@ pmap_getport (address, program, version,
   struct pmap parms;
 
   address->sin_port = htons (PMAPPORT);
-  client = clntudp_bufcreate (address, PMAPPROG,
+  if (protocol == IPPROTO_TCP)
+    {
+      // glibc does this:
+      ///* Don't need a reserved port to get ports from the portmapper.  */
+      //socket = __get_socket(address); // does socket(TCP),bind(),connect(address)
+      //if (_socket != -1)
+      //  closeit = true;
+      // do we need/want to do the same?
+      client = clnttcp_create (address, PMAPPROG,
+	      PMAPVERS, &_socket, RPCSMALLMSGSIZE, RPCSMALLMSGSIZE);
+    }
+  else
+    client = clntudp_bufcreate (address, PMAPPROG,
 	      PMAPVERS, timeout, &_socket, RPCSMALLMSGSIZE, RPCSMALLMSGSIZE);
   if (client != (CLIENT *) NULL)
     {
