--- busybox-1.13.2/uClibc-0.9.30.1/libc/sysdeps/linux/common/mknod.c	2008-10-28 14:05:24.000000000 +0100
+++ busybox-1.13.2/uClibc-0.9.30.1-mknod/libc/sysdeps/linux/common/mknod.c	2012-12-13 18:45:09.463515633 +0100
@@ -19,10 +19,6 @@ static __inline__ _syscall3(int, __sysca
 
 int mknod(const char *path, mode_t mode, dev_t dev)
 {
-	/* We must convert the dev_t value to a __kernel_dev_t */
-	__kernel_dev_t k_dev;
-
-	k_dev = ((major(dev) & 0xff) << 8) | (minor(dev) & 0xff);
-	return __syscall_mknod(path, mode, k_dev);
+	return __syscall_mknod(path, mode, (__kernel_dev_t)dev);
 }
 libc_hidden_def(mknod)
